Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,321 @@
+def gradient (f : F → 𝕜) (x : F) : F := (toDual 𝕜 F).symm (fderiv 𝕜 f x)
+
+local notation "∇" => gradient
+
+variable {s : Set F} {L : Filter F}
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt {s : Set F} :
+    HasGradientWithinAt f f' s x ↔ HasFDerivWithinAt f (toDual 𝕜 F f') s x := Iff.rfl
+
+theorem HasGradientAt_iff_HasFDerivAt :
+    HasGradientAt f f' x ↔ HasFDerivAt f (toDual 𝕜 F f') x := Iff.rfl
+
+theorem HasGradientAt.hasFDerivAt {frechet : F →L[𝕜] 𝕜}
+    (h : HasGradientAt f ((toDual 𝕜 F).symm frechet) x) :
+    HasFDerivAt f frechet x := by
+  rw [← (toDual 𝕜 F).apply_symm_apply frechet]
+  exact HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasFDerivAt.hasGradientAt {frechet : F →L[𝕜] 𝕜} (h : HasFDerivAt f frechet x) :
+    HasGradientAt f ((toDual 𝕜 F).symm frechet) x := by
+  rw [← (toDual 𝕜 F).apply_symm_apply frechet] at h
+  exact HasGradientAt_iff_HasFDerivAt.mpr h
+theorem Gradient_zero_of_not_differentiableAt (h : ¬DifferentiableAt 𝕜 f x) : ∇ f x = 0 := by
+  have : (toDual 𝕜 F).symm 0 = 0 := by simp only [map_zero]
+  rw [← fderiv_zero_of_not_differentiableAt h] at this
+  exact this
+
+local notation "⟪" x ", " y "⟫" => @inner 𝕜 _ _ x y
+
+section GradientUniqueness
+
+variable {gradf gradg : F}
+
+theorem HasGradientAt.unique (hf : HasGradientAt f gradf x) (hg : HasGradientAt f gradg x) :
+    gradf = gradg := (toDual 𝕜 F).injective
+  ((HasGradientAt_iff_HasFDerivAt.mp hf).unique (HasGradientAt_iff_HasFDerivAt.mp hg))
+
+end GradientUniqueness
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt' {s : Set F} {frechet : F →L[𝕜] 𝕜} :
+    HasGradientWithinAt f ((toDual 𝕜 F).symm frechet) s x ↔ HasFDerivWithinAt f frechet s x := by
+  conv_rhs => rw [← (toDual 𝕜 F).apply_symm_apply frechet]
+
+theorem DifferentiableAt.hasGradientAt (h : DifferentiableAt 𝕜 f x) :
+    HasGradientAt f (∇ f x) x := by
+  rw [HasGradientAt_iff_HasFDerivAt, gradient, (toDual 𝕜 F).apply_symm_apply (fderiv 𝕜 f x)]
+  exact h.hasFDerivAt
+
+theorem HasGradientAt.differentiableAt (h : HasGradientAt f f' x) :
+    DifferentiableAt 𝕜 f x := by use ((toDual 𝕜 F) f'); apply HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasGradientWithinAt.differentiableWithinAt (h : HasGradientWithinAt f f' s x) :
+    DifferentiableWithinAt 𝕜 f s x := HasFDerivWithinAt.differentiableWithinAt h
+
+@[simp]
+theorem hasGradientWithinAt_univ : HasGradientWithinAt f f' univ x ↔ HasGradientAt f f' x := by
+  rw [HasGradientWithinAt_iff_HasFDerivWithinAt, HasGradientAt_iff_HasFDerivAt]
+  rw [hasFDerivWithinAt_univ]
+
+theorem DifferentiableOn.hasGradientAt (h : DifferentiableOn 𝕜 f s) (hs : s ∈ 𝓝 x) :
+    HasGradientAt f (∇ f x) x := (h.hasFDerivAt hs).hasGradientAt
+
+theorem HasGradientAt.gradient (h : HasGradientAt f f' x) : ∇ f x = f' :=
+  h.differentiableAt.hasGradientAt.unique h
+
+theorem gradient_eq {f' : F → F} (h : ∀ x, HasGradientAt f (f' x) x) : ∇ f = f' :=
+  funext fun x => (h x).gradient
+
```

Comment:
Can you open a section here, named `OneDimension` or something like that?

Reply:
I don't think you have opened the section at the right place. It should probably be before the `variable` line, as all theorems talking about `g` will be one-dimensional theorems.